### DISCUSSION TITLE
Multiple rest client instances

### DISCUSSION THREAD

**Discussion by ichrist-gr** [quarkusio/quarkus#28785] [Q&A] [answered]
Hello all,
I have an application that sends loads of http requests in another service.  Performance concerns us, so we decided to use vertx multiple verticle instances, to perform these requests. But as far as i understand, all these verticle instances (10) use the same rest client instance, so we see no performance benefit. 

How can i overcome this issue and **inject multiple rest client instances (one per each verticle instance)**

Thank you very much.

[a comment by an automated account is not shown]

**Comment by ichrist-gr** [comment]
I should also include that we are using the **reactive** libraries for resteasy and rest client and we have registered the config for connections:
connection-pool-size: 10

**Comment by geoand** [comment, marked as answer]
Hi,

> Performance concerns us, so we decided to use vertx multiple verticle instances, to perform these requests

Can you describe your performance concern and what your thought process on using multiple verticle instances was? 
I want to understand the root cause before trying to give an answer.

Thanks

**Reply by cescoffier** [reply]
Multi-verticle won't help much as it is shared. 
You can:
- increase the number of concurrent connections to your endpoint. 
- use the web client directly and instantiate it once per verticle

**Reply by ichrist-gr** [reply]
Thank you for answering

we have a case which:
1) Retrieve lists of events every x seconds and call some external services.
2)  One of the external services, composes of 4 rest calls.

So assigning an event loop for each event to perform 4 rest calls might causes some performance issues.
So if we deploy multiple instances of the verticle that connects with the rest client will perform better (opinion).

**Reply by geoand** [reply]
Actually, what you describe is a very good case for using the event loop concept (or more accurately, AsyncIO in general). As the Reactive REST Client will not block the thread, there is no reason to worry about performance.
What do you need to look out for however is what @cescoffier says - you may need to increase the concurrent connection number as by default the client only allows a certain about of inflight requests (5 I think).

**Reply by ichrist-gr** [reply]
1) We have configured connection-pool-size: 10 but no better results
2) I will try to instantiate the vert.x web client per verticle and check the results

Thank you both for your answers i will come back with the results

**Reply by geoand** [reply]
Are you using a `Uni` as a return type? If you don't, the current thread will be blocked (even though the IO will be non blocking)

**Reply by ichrist-gr** [reply]
Yes we are using the reactive types in each api call and also the reactive libraries 

Short description:
We receive a list of 50 events with a scheduler.
So we break this list into a list of 10 events. These 10 events via multi.create.fromiterable() are processed and sent in these 2 external systems, so our thought was that if we scale up the service with the 4 rest calls, each event will be assigned in a different event loop (that's why we defined the instance of the verticle to 10) and when the results show up, they will be emitted. Then the next 10 list of events will be processed and again the will follow the same journey

 **So scaling up the verticle with the 4 api calls will not give any benefit?**

Below there is a code example of the orchestrator on how the list with 50 events is separated and emits the results:

```java
private Uni<List<TaskDetails>> orchestrateTasksBatch(TaskQueue taskQueue) {
        return Multi.createFrom()
                .iterable(taskQueue.getTaskDetails())
                .group()
                .intoLists()
                .of(10)
                .onItem()
                .transformToMultiAndMerge(tasks -> handleTaskSubList(taskQueue, tasks))
                .collect()
                .asList();
    }

    private Multi<TaskDetails> handleTaskSubList(TaskQueue taskQueue, List<TaskDetails> tasks) {
        AtomicInteger taskCounter = new AtomicInteger();

        return Multi.createFrom()
                .emitter(emitter ->
                        Multi.createFrom()
                                .iterable(tasks)
                                .subscribe()
                                .with(task -> handleTask(taskQueue, taskCounter, emitter, task)));
    }

    private void handleTask(TaskQueue taskQueue,
                            AtomicInteger taskCounter,
                            MultiEmitter<? super TaskDetails> emitter,
                            TaskDetails task) {
        taskService.processTask(task, taskQueue)
                .onItem()
                .invoke(taskCounter::incrementAndGet)
                .subscribe()
                .with(taskResult -> {
                    emitter.emit(taskResult);
                    if (taskCounter.get() == 10) {
                        emitter.complete();
                    }
                });
    }
```

**Reply by ichrist-gr** [reply]
The taskService.processTask method will interact with the 2 external systems.

Important note is that it does the job as expected, but we would like to make it perform better

**Reply by geoand** [reply]
I am still unclear of where you think the performance problems are

**Reply by ichrist-gr** [reply]
Sorry for confusing you george. 
The way it is right now, each batch of 50 events will be finished in 1 minute. 
So when invoking the service for each event (taskService.processTask)  it's journey demands 4 api calls.   
So we assume that if we wrap the taskService.processTask into a verticle and scale up this verticle with 10 instances, each event will be processed in different event loops. 
Will this assumption be an actual performance increase or am i getting something wrong?

In fact, we want to achieve a parallel execution for the event sublists.

Thank you very much

**Reply by geoand** [reply]
@cescoffier can tell you more for sure, but AFAIK, it won't make a difference. If I were you, I would increase `quarkus.rest-client.myclient.connection-pool-size`.
If that does not increase the performance of the system, you'll probably have to introduce tracing to figure out what's going on.

**Reply by cescoffier** [reply]
If you use a single rest client - you won't see any benefit, because it will use a single event loop.

If you use a web client, then, using verticles could work. It will depends on the number of core. 10 verticles would require 10 cores, or you will overcommit. 

As said by @geoand, I would recommend increasing the number of connection and keep thing simple. Observe and measure before going to direction that may not bring any benefits.

**Reply by ichrist-gr** [reply]
Thank you for your time guys!!

**Comment by ichrist-gr** [comment]
Returned with some results. Indeed guys playing around with the connection pool size and benchmarking every time, ended up in a performance increase. (Other's system gateway started rejecting me because of heavy load😇).

Thank you very much!!!

**Reply by geoand** [reply]
Great to hear that the Quarkus bottleneck was eliminated :)
